monitoring/check_collectd_mlab_test.py

```diff
@@ -0,0 +1,301 @@
+#!/usr/bin/python
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for check_collectd_mlab check."""
+
+import os
+import socket
+import time
+import unittest2 as unittest
+
+# Third-party modules.
+import mock
+
+# Module under test.
+import check_collectd_mlab
+
+# R0904: Too many public methods. Hard to avoid for unit tests.
+# pylint: disable=R0904
+
+
+class MLabNagiosSocketTests(unittest.TestCase):
+
+  def setUp(self):
+    self._testdata_dir = os.path.join(
+        os.path.dirname(check_collectd_mlab.__file__), 'testdata')
+
+  def testunit_sock_sendcmd_RETURNS_successfully(self):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.recv.side_effect = list('1 default reply\n')
```

Comment:
My understanding of the concern here is that the test code is unintuitive. And, my concern is that we cannot use the unittest.Mock interface in the way you suggest for the current implementation of  `check_collectd_mlab.sock_readline`.

So, I created a `FakeSocketIO` class that is socket-like, supporting `recv` and `send` methods (those used by functions in check_collectd_mlab). Think `StringIO` for sockets. All data to-be-received can be declared initially as a single string, and all sent data can be queried after the test.

I believe this interface is more intuitive because the fake object behaves like an actual socket for the tests code and setup for the FakeSocketIO is straight forward.

How does it look? 301873883cb1227375a48ea6b41b3d57a9c56fa3

So, the arc of this code has been from the FakeCollectd (which while functionally complete was a bit complex) to Mock (which is unintuitive) to FakeSocketIO (which I hope strikes a balance).